File: node/src/runtime/credential-vault.ts
export interface CredentialVault {
  get(key: string): Promise<string | undefined>
}

export class EnvCredentialVault implements CredentialVault {
  async get(key: string): Promise<string | undefined> {
    return process.env[key]
  }
}

export class InMemoryCredentialVault implements CredentialVault {
  private store: Map<string, string>

  constructor(init?: Record<string, string>) {
    this.store = new Map(Object.entries(init ?? {}))
  }

  set(key: string, value: string): this {
    this.store.set(key, value)
    return this
  }

  async get(key: string): Promise<string | undefined> {
    return this.store.get(key)
  }
}

/** Tries each vault in order, returning the first defined value. */
export class ChainedCredentialVault implements CredentialVault {
  constructor(private readonly vaults: CredentialVault[]) {}

  async get(key: string): Promise<string | undefined> {
    for (const vault of this.vaults) {
      const val = await vault.get(key)
      if (val !== undefined) return val
    }
    return undefined
  }
}

File: rust/src/runtime/credential_vault.rs
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

/// Read-only credential abstraction. Credentials never enter model context or session log.
#[async_trait]
pub trait CredentialVault: Send + Sync {
    async fn get(&self, key: &str) -> Option<String>;
}

/// Reads credentials from environment variables.
pub struct EnvCredentialVault;

#[async_trait]
impl CredentialVault for EnvCredentialVault {
    async fn get(&self, key: &str) -> Option<String> {
        std::env::var(key).ok()
    }
}

/// In-memory credential store, useful for tests and embedded deployments.
pub struct InMemoryCredentialVault {
    store: Arc<RwLock<HashMap<String, String>>>,
}

impl Default for InMemoryCredentialVault {
    fn default() -> Self {
        Self { store: Arc::new(RwLock::new(HashMap::new())) }
    }
}

impl InMemoryCredentialVault {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn set(&self, key: impl Into<String>, value: impl Into<String>) {
        self.store.write().await.insert(key.into(), value.into());
    }
}

#[async_trait]
impl CredentialVault for InMemoryCredentialVault {
    async fn get(&self, key: &str) -> Option<String> {
        self.store.read().await.get(key).cloned()
    }
}

/// Tries each vault in order, returning the first non-`None` result.
pub struct ChainedCredentialVault {
    vaults: Vec<Box<dyn CredentialVault>>,
}

impl ChainedCredentialVault {
    pub fn new(vaults: Vec<Box<dyn CredentialVault>>) -> Self {
        Self { vaults }
    }
}

#[async_trait]
impl CredentialVault for ChainedCredentialVault {
    async fn get(&self, key: &str) -> Option<String> {
        for vault in &self.vaults {
            if let Some(v) = vault.get(key).await {
                return Some(v);
            }
        }
        None
    }
}
